api/envoy/config/endpoint/v3/endpoint_components.proto

```diff
@@ -44,6 +44,17 @@ message Endpoint {
     // to a non-empty value allows overriding the cluster level configuration for a specific
     // endpoint.
     string hostname = 2;
+
+    // Optional alternative health check host address.
+    //
+    // .. attention::
+    //
+    //   The form of host address depends on the given cluster type. For STATIC or EDS,
+    //   it is expected to be a direct IP address (or something resolvable by the
+    //   specified :ref:`resolver <envoy_v3_api_field_config.core.v3.SocketAddress.resolver_name>`
+    //   in the Address). For LOGICAL or STRICT DNS, it is expected to be hostname,
+    //   and will be resolved via DNS.
```

Comment:
This should have more testing. Please add an actual integration test in https://github.com/envoyproxy/envoy/blob/main/test/integration/health_check_integration_test.cc, and also test logical/strict DNS, etc.

Reply:
The other review comments I have addressed and tested locally.

While I was trying to add the tests for this comment, I found that in order to support hostname that resolvable via DNS
We might need to add a health check address resolve target here and similarly in logic_dns_cluster.cc
https://github.com/envoyproxy/envoy/blob/main/source/common/upstream/strict_dns_cluster.cc#L34

If the resolution comes back with multiple addresses, we ends up with multiple endpoints with different health check addresse  or the endpoints equals to a matrix of endpoint addresses with endpoint health check addresses.

Our team's use case is to allow the integration of envoy health check with our centralized health checker which has a fixed ip address.

Can we limit the address to be just a direct IP address?

Or if you prefer it to be generalized and it can be both hostname and direct IP address?

Reply:
I would limit and it can always be relaxed later. IMO a fixed IP is not super user friendly, so you might want to support a hostname which will use only a single address or something like that, but up to you as long as it is clearly documented.

Reply:
Thanks, I chose to limit it.
I'll do a PR for a custom health check extension which requires this PR.
Then after that I'll do another PR to relax it to be both hostname and IP.

Hopefully this is acceptable for you.

health_check_integration_test is using ConfigHelper::buildStaticCluster for building the test cluster
https://github.com/envoyproxy/envoy/blob/main/test/integration/health_check_integration_test.cc#L82

So I did the change below, then the new code gets tested.
https://github.com/envoyproxy/envoy/pull/23000/files#diff-ce04c82dd2b166c6df41e514e7c253ceeb276f7db1d186ddc16f8f1841ad9496R498